// snabbdom/src/package/h.ts
import { vnode, VNode, VNodeData } from './vnode'
import * as is from './is'

export type VNodes = VNode[]
export type VNodeChildElement = VNode | string | number | undefined | null
export type ArrayOrElement<T> = T | T[]
export type VNodeChildren = ArrayOrElement<VNodeChildElement>

function addNS (data: any, children: VNodes | undefined, sel: string | undefined): void {
  data.ns = 'http://www.w3.org/2000/svg'
  if (sel !== 'foreignObject' && children !== undefined) {
    for (let i = 0; i < children.length; ++i) {
      const childData = children[i].data
      if (childData !== undefined) {
        addNS(childData, (children[i] as VNode).children as VNodes, children[i].sel)
      }
    }
  }
}


// TS 中方法重构的方式
/**
 * 
 * @param sel 是创建虚拟 Dom 的 标签类型及其 id class 选择器的 name
 * @param data 是虚拟 DOM 的模块配置比如 style , arrtibutes(属性) props class 等, 对象传入
 * @param VNodeChildren DOM 元素的子节点（可以是 HTML 元素 也可以是 #text 节点）
 * 
 * 
 */
export function h (sel: string): VNode
export function h (sel: string, data: VNodeData | null): VNode
export function h (sel: string, children: VNodeChildren): VNode
export function h (sel: string, data: VNodeData | null, children: VNodeChildren): VNode
export function h (sel: any, b?: any, c?: any): VNode {
  var data: VNodeData = {}
  var children: any 
  var text: any
  var i: number
  //  如果 c 不等于 空时 
  if (c !== undefined) {
    if (b !== null) {
      data = b
    }
    // 判断为数组时, c === VNodeChildren
    if (is.array(c)) {
      children = c
      // 判断为原始值 number string boolean 时
    } else if (is.primitive(c)) {
      text = c
      // 如果 c 是一个 Vnode 时
    } else if (c && c.sel) {
      children = [c]
    }
  } else if (b !== undefined && b !== null) {
    if (is.array(b)) {
      children = b
    } else if (is.primitive(b)) {
      text = b
    } else if (b && b.sel) {
      children = [b]
    } else { data = b }
  }
  if (children !== undefined) {
    for (i = 0; i < children.length; ++i) {
      if (is.primitive(children[i])) children[i] = vnode(undefined, undefined, undefined, children[i], undefined)
    }
  }
  if (
    sel[0] === 's' && sel[1] === 'v' && sel[2] === 'g' &&
    (sel.length === 3 || sel[3] === '.' || sel[3] === '#')
  ) {
    addNS(data, children, sel)
  }
  return vnode(sel, data, children, text, undefined)
};
